File: lc_soln/1353_maximum_number_of_events_that_can_be_attended/solution.cpp
#include <algorithm>
#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
    int maxEvents(vector<vector<int>>& events) {
        ranges::sort(events, [](const vector<int>& a, const vector<int>& b) {
            return a[0] < b[0];
        });
        priority_queue<int, vector<int>, greater<int>> pq;
        int result = 0;
        int i = 0;
        int n = events.size();
        int day = events[0][0];
        while (i < n || !pq.empty()) {
            while (!pq.empty() && pq.top() < day) {
                pq.pop();
            }
            while (i < n && events[i][0] == day) {
                pq.push(events[i][1]);
                ++i;
            }
            if (!pq.empty()) {
                pq.pop();
                ++result;
            }
            ++day;
        }
        return result;
    }
};

File: lc_soln/1353_maximum_number_of_events_that_can_be_attended/solution.py
import heapq
from typing import List


class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        events = sorted(events, key=lambda x: x[0])
        h = []
        day = 1
        i = 0
        n = len(events)
        result = 0
        while i < n or h:
            while h and h[0] < day:
                heapq.heappop(h)

            while i < n and events[i][0] == day:
                heapq.heappush(h, events[i][1])
                i += 1

            if h:
                heapq.heappop(h)
                result += 1

            day += 1
        return result
